README.md
# Визуализация атома

[![License](https://img.shields.io/badge/license-MIT-informational)](LICENSE)
[![GitHub release](https://img.shields.io/github/release/YuliaMakarova/atom-visualization.svg)](https://github.com/YuliaMakarova/atom-visualization/releases/latest)

![](https://github.com/YuliaMakarova/atom-visualization/raw/master/gif/example.gif)

## Содержание

- [Описание](#описание)
- [Требования](#требования)
- [Установка и запуск](#установка-и-запуск)

## Описание

Данный проект представляет собой простую визуализацию атома с использованием библиотеки Pygame и OpenGL. В атоме присутствуют ядро и вращающиеся вокруг него электроны. Электроны движутся по орбитам, создавая динамичное представление атомной структуры.

## Требования
- Python 3.x
- Библиотека Pygame
- Библиотека OpenGL

## Установка и запуск

1. Склонируйте репозиторий на свой локальный компьютер:
   ```bash
   git clone https://github.com/YuliaMakarova/atom-visualization.git
2. Перейдите в каталог проекта:
    ```bash
    cd atom-visualization
3. Установите необходимые зависимости:
    ```bash
    pip install -r requirements.txt

4. Запустите визуализацию атома:
    ```bash
    python main.py



CHANGELOG.md
# Changelog

## [1.0.0] - 2023-11-17

### Новые функции

- Реализована отрисовка ядра с использованием OpenGL.
- Добавлена функциональность для отрисовки электронов в случайных орбитах вокруг ядра.
- Настроено освещение и материалы для реалистичного отображения атома.
- Установлен базовый режим отображения Pygame для отображения модели молекулы.
- Включены параметры для количества электронов, радиуса ядра и свойств электронов.
- Использованы библиотеки PyOpenGL и Pygame для трехмерной графики и управления окном.
- Созданы функции для инициализации, отрисовки ядра и отрисовки электронов.
- Определены константы для размера окна и границ.

[1.0.0]: https://github.com/YuliaMakarova/atom-visualization/releases/tag/v1.0.0

tests/test_main.py
import sys
import os
import random
import unittest
import pygame
from pygame.locals import DOUBLEBUF, OPENGL
from OpenGL.GL import glTranslatef
from OpenGL.GLU import gluPerspective
# Получаем путь к текущему каталогу скрипта
current_dir = os.path.dirname(os.path.abspath(__file__))
# Добавляем путь к родительскому каталогу
sys.path.append(os.path.join(current_dir, '..'))
from main import draw_electron, draw_nucleus


class TestDrawFunctions(unittest.TestCase):
    """
    Тестирование функций отрисовки электрона и ядра.
    """

    def setUp(self):
        """
        Настройка окружения для тестов.
        """
        pygame.init()
        self.display = (1200, 800)
        pygame.display.set_mode(self.display, DOUBLEBUF | OPENGL)
        gluPerspective(45, (self.display[0] / self.display[1]), 0.1, 50.0)
        glTranslatef(0.0, 0.0, -5)

    def tearDown(self):
        """
        Очистка окружения после тестов.
        """
        pygame.quit()

    def test_draw_electron(self):
        """
        Тестирование функции отрисовки электрона.
        """
        electron_angles = random.uniform(0, 360)
        electron_orbit_radii = random.uniform(1.0, 2.5)
        electron_orbit_angles = random.uniform(0, 360)
        electron_diffuse_colors = (0.0, 0.0, 1.0, 1.0)
        self.assertIsNone(draw_electron(electron_angles,
                                        electron_orbit_radii,
                                        electron_orbit_angles,
                                        electron_diffuse_colors))

    def test_draw_nucleus(self):
        """
        Тестирование функции отрисовки ядра.
        """
        self.assertIsNone(draw_nucleus())


main.py
import pygame
from pygame.locals import DOUBLEBUF, OPENGL
from OpenGL.GL import glClear, GL_COLOR_BUFFER_BIT, GL_DEPTH_BUFFER_BIT, \
    glEnable, GL_LIGHTING, GL_LIGHT0, GL_POSITION, GL_SPOT_DIRECTION, GL_DIFFUSE, GL_SPECULAR, GL_SHININESS, \
    glLoadIdentity, glTranslatef, glRotatef, glPushMatrix, glPopMatrix, glClearDepth, glDepthFunc, GL_LESS, \
    glMaterialfv, GL_FRONT_AND_BACK, GL_EMISSION, glLightfv, GL_DEPTH_TEST, glGetDoublev, GL_MODELVIEW_MATRIX
from OpenGL.GLU import gluNewQuadric, gluSphere, gluPerspective
import random

# Количество электронов
NUM_ELECTRONS = 6

# Параметры ядра
NUCLEUS_RADIUS = 0.8
NUCLEUS_COLOR = (1.0, 0.0, 0.0, 1.0)

# Параметры электрона
ELECTRON_RADIUS = 0.2
ELECTRON_SPEED = 2.0

# Цвета для диффузного отражения электронов
ELECTRON_DIFFUSE_COLORS = [(0.0, 0.0, 1.0, 1.0) for _ in range(NUM_ELECTRONS)]

# Углы поворота электронов
ELECTRON_ANGLES = [random.uniform(0, 360) for _ in range(NUM_ELECTRONS)]
# Радиусы орбит электронов
ELECTRON_ORBIT_RADII = [random.uniform(1.0, 2.5) for _ in range(NUM_ELECTRONS)]
# Углы наклона орбит электронов
ELECTRON_ORBIT_ANGLES = [random.uniform(0, 360) for _ in range(NUM_ELECTRONS)]

# Константы для размеров окна
WINDOW_WIDTH = 1200
WINDOW_HEIGHT = 800

# Границы окна
WINDOW_X_LIMIT = WINDOW_WIDTH / 2
WINDOW_Y_LIMIT = WINDOW_HEIGHT / 2


def init():
    """
    Инициализирует OpenGL, устанавливая освещение и материалы для отображения трехмерной модели атома.

    Освещение:
    - Включает освещение и источник света GL_LIGHT0.
    - Устанавливает позицию и направление света.

    Основные параметры:
    - position: Позиция светового источника в пространстве (x, y, z, w).
    - spot_direction: Направление светового пятна.

    Трансформации:
    - Устанавливает матрицу проекции, перспективу и перемещение взгляда камеры.

    Глубина:
    - Включает тест глубины и устанавливает параметры теста.

    Материалы:
    - Устанавливает диффузное отражение, зеркальность и блеск материала атома.

    :return: Нет возвращаемых значений.
    """
    glEnable(GL_LIGHTING)
    glEnable(GL_LIGHT0)
    glEnable(GL_DEPTH_TEST)
    glEnable(GL_COLOR_BUFFER_BIT)

    # Позиция светового источника
    position = [0.0, 0.0, 5.0, 1.0]
    # Направление светового пятна
    spot_direction = [0.0, 0.0, -1.0]

    glLoadIdentity()
    gluPerspective(45, (WINDOW_WIDTH / WINDOW_HEIGHT), 0.1, 50.0)
    glTranslatef(0.0, 0.0, -5)

    # Настройка теста глубины
    glClearDepth(1.0)
    glDepthFunc(GL_LESS)

    # Установка позиции и направления светового источника
    glLightfv(GL_LIGHT0, GL_POSITION, position)
    glLightfv(GL_LIGHT0, GL_SPOT_DIRECTION, spot_direction)

    # Установка свойств материала атома
    material_diffuse = [0.8, 0.8, 0.8, 1.0]
    material_specular = [1.0, 1.0, 1.0, 1.0]
    material_shininess = [50.0]

    glMaterialfv(GL_FRONT_AND_BACK, GL_DIFFUSE, material_diffuse)
    glMaterialfv(GL_FRONT_AND_BACK, GL_SPECULAR, material_specular)
    glMaterialfv(GL_FRONT_AND_BACK, GL_SHININESS, material_shininess)


def draw_nucleus():
    """
    Отрисовывает ядро атома с эмиссионным свечением.

    Использует библиотеку OpenGL для создания сферы, представляющей ядро атома.
    Устанавливает цвет эмиссии для создания свечения ядра.

    :return: Нет возвращаемых значений.
    """
    # Устанавливает цвет эмиссии для свечения ядра
    glMaterialfv(GL_FRONT_AND_BACK, GL_EMISSION, NUCLEUS_COLOR)

    # Создает квадрику для отрисовки сферы
    quadric = gluNewQuadric()

    # Рисует сферу с указанными параметрами
    gluSphere(quadric, NUCLEUS_RADIUS, 20, 20)

    # Сбрасывает цвет эмиссии после отрисовки
    glMaterialfv(GL_FRONT_AND_BACK, GL_EMISSION, [0.0, 0.0, 0.0, 1.0])


def draw_electron(angle, orbit_radius, orbit_angle, electron_diffuse_color):
    """
    Отрисовывает электрон, представленный сферой, с учетом его орбитального положения и цвета диффузного отражения.

    :param angle: Угол поворота электрона вокруг своей оси.
    :param orbit_radius: Радиус орбиты электрона.
    :param orbit_angle: Угол наклона орбиты электрона.
    :param electron_diffuse_color: Цвет диффузного отражения для электрона.
    :return: Нет возвращаемых значений.
    """
    # Сохраняет текущую матрицу модели-вида
    glPushMatrix()

    # Поворачивает и транслирует для установки положения электрона
    glRotatef(orbit_angle, 0.0, 0.0, 1.0)
    glRotatef(angle, 0.0, 1.0, 0.0)
    glTranslatef(orbit_radius, 0.0, 0.0)

    # Проверяет границы перед движением
    new_pos = glGetDoublev(GL_MODELVIEW_MATRIX)
    electron_x = new_pos[3][0]
    electron_y = new_pos[3][1]

    if abs(electron_x) > WINDOW_X_LIMIT or abs(electron_y) > WINDOW_Y_LIMIT:
        # Возвращает предыдущую матрицу, чтобы отменить движение
        glPopMatrix()
    else:
        # Устанавливает цвет диффузного отражения для электрона
        glMaterialfv(GL_FRONT_AND_BACK, GL_DIFFUSE, electron_diffuse_color)

        # Создает квадрику для отрисовки сферы
        quadric = gluNewQuadric()

        # Рисует сферу с указанными параметрами
        gluSphere(quadric, ELECTRON_RADIUS, 10, 10)

    # Восстанавливает предыдущую матрицу модели-вида
    glPopMatrix()


def draw():
    """
    Осуществляет отрисовку сцены, включая ядро и электроны.

    :return: Нет возвращаемых значений.
    """
    # Очищает буферы цвета и глубины
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)

    # Устанавливает единичную матрицу проекции
    glLoadIdentity()
    gluPerspective(45, (WINDOW_WIDTH / WINDOW_HEIGHT), 0.1, 50.0)

    # Транслирует вглубь экрана
    glTranslatef(0.0, 0.0, -5)

    # Отрисовывает ядро
    draw_nucleus()

    # Отрисовывает электроны
    for i in range(NUM_ELECTRONS):
        draw_electron(
            ELECTRON_ANGLES[i],
            ELECTRON_ORBIT_RADII[i],
            ELECTRON_ORBIT_ANGLES[i],
            ELECTRON_DIFFUSE_COLORS[i]
        )
        ELECTRON_ANGLES[i] += ELECTRON_SPEED

    # Обновляет дисплей и ожидает 10 миллисекунд для контроля частоты кадров
    pygame.display.flip()
    pygame.time.wait(10)


def main():
    """
    Основная функция программы, инициализирует Pygame, устанавливает режим отображения,
    и запускает бесконечный цикл для отрисовки сцены.

    :return: Нет возвращаемых значений.
    """
    # Инициализирует Pygame
    pygame.init()

    # Устанавливает размеры окна
    display = (WINDOW_WIDTH, WINDOW_HEIGHT)

    # Устанавливает режим отображения с двойным буфером и использованием OpenGL
    pygame.display.set_mode(display, DOUBLEBUF | OPENGL)

    # Инициализирует OpenGL
    init()

    # Бесконечный цикл для отрисовки сцены
    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                # Завершает программу при закрытии окна
                pygame.quit()
                quit()

        # Осуществляет отрисовку сцены
        draw()


if __name__ == "__main__":
    main()
